### PR TITLE
DNM WIP 20220624 zipper loadable plugins

### PR DESCRIPTION
**Pull request by kalebskeithley**
Adds (shared) libs:
rgw_sal_rados (librgw_sal_rados) rados plugin
rgw_sal_d3n (librgw_sal_d3n) d3n plugin
rgw_sal_motr (librgw_sal_mtr) motr plugin
rgw_sal_dbstore (librgw_sal_dbstore) dbstore plugin
rados_sal library (librados_sal), below the line shlib for rgw_sal_rados and rgw_sal_d3n

future:
rgw_sal_{rados,d3n,motr,dbstore} sources move to their own respective subdirs.
rados_sal sources move to a subdir
dlopen/dlsym the plugin(s) based on config instead of hard-coded

I'm not wedded to the names so if you want different names for the new
libs I'm open to changing them.

With this refactoring we have the remaining undefined ref issues across the line (above line, below line)

13 non RGWSI* (note IIUC the rgw::notify* refs will be resolved by forthcoming tracing refactoring which will eliminate four more issues:)

/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWBucketCtl::read_buckets_stats(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, RGWBucketEnt, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, RGWBucketEnt> > >&, optional_yield, DoutPrefixProvider const*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWHandler_REST_PSNotifs_S3::create_delete_op()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWHandler_REST_PSNotifs_S3::create_get_op()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWHandler_REST_PSNotifs_S3::create_put_op()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `rgw_init_ioctx(DoutPrefixProvider const*, librados::v14_2_0::Rados*, rgw_pool const&, librados::v14_2_0::IoCtx&, bool, bool)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWMetaSyncStatusManager::init(DoutPrefixProvider const*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `rgw::notify::publish_commit(rgw::sal::Object*, unsigned long, std::chrono::time_point<ceph::real_clock, std::chrono::duration<unsigned long, std::ratio<1l, 1000000000l> > > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw::notify::EventType, rgw::notify::reservation_t&, DoutPrefixProvider const*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `rgw::notify::publish_reserve(DoutPrefixProvider const*, rgw::notify::EventType, rgw::notify::reservation_t&, RGWObjTags const*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `rgw::notify::reservation_t::~reservation_t()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWRemoteMetaLog::read_sync_status(DoutPrefixProvider const*, rgw_meta_sync_status*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `vtable for RGWHandler_Bucket'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `vtable for RGWHandler_Log'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `vtable for RGWHandler_REST_PSTopic_AWS'

and 23 RGWSI* issues:

/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWMetadataHandlerPut_SObj::RGWMetadataHandlerPut_SObj(RGWMetadataHandler_GenericMetaBE*, RGWSI_MetaBackend_Handler::Op*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, RGWMetadataObject*, RGWObjVersionTracker&, optional_yield, RGWMDLogSyncType, bool)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Cls::MFA::check_mfa(DoutPrefixProvider const*, rgw_user const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, optional_yield)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_MetaBackend_Handler::Op_ManagedCtx::Op_ManagedCtx(RGWSI_MetaBackend_Handler*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_MetaBackend::PutParams::~PutParams()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Role_RADOS::get_be_handler()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::get_obj(RGWSysObjectCtx&, rgw_raw_obj const&)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::get_zone_svc()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::init_obj_ctx()'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::ROp::read(DoutPrefixProvider const*, long, long, ceph::buffer::v15_2_0::list*, optional_yield)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::ROp::ROp(RGWSI_SysObj::Obj&)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::ROp::stat(optional_yield, DoutPrefixProvider const*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::WNOp::notify(DoutPrefixProvider const*, ceph::buffer::v15_2_0::list&, unsigned long, ceph::buffer::v15_2_0::list*, optional_yield)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::WOp::remove(DoutPrefixProvider const*, optional_yield)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_SysObj::Obj::WOp::write(DoutPrefixProvider const*, ceph::buffer::v15_2_0::list&, optional_yield)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::find_zone_id_by_name(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, rgw_zone_id*)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::find_zone(rgw_zone_id const&, RGWZone**)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_realm() const'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_sync_policy_handler(std::optional<rgw_zone_id>) const'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_zonegroup() const'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::get_zone_params() const'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::list_realms(DoutPrefixProvider const*, std::__cxx11::list<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::list_zonegroups(DoutPrefixProvider const*, std::__cxx11::list<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&)'
/usr/bin/ld: lib/libradosgw.so.2.0.0: undefined reference to `RGWSI_Zone::list_zones(DoutPrefixProvider const*, std::__cxx11::list<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&)'

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_lua.cc [hunk 1]
```diff
@@ -5,11 +5,6 @@
 #include "rgw_lua_utils.h"
 #include "rgw_sal_rados.h"
 #include "rgw_lua.h"
-#ifdef WITH_RADOSGW_LUA_PACKAGES
```
Comment (yuvalif) [inline src/rgw/rgw_lua.cc:8]: why was this removed?
this is used to support systems that do not have luarocks (IIRC, opensuse, freebsd)
also, this could be used by downstream packagers that don't want to allow for any external lua libs (for security reasons)

File: src/rgw/rgw_sal_rados.cc [hunk 2]
```diff
@@ -3043,6 +3044,171 @@ int RadosLuaScriptManager::del(const DoutPrefixProvider* dpp, optional_yield y,
...
+      return ret;
+    }
+
+    packages.merge(packages_chunk);
+  }
+
+  return 0;
+}
+
+int RadosLuaManager::install_packages(const DoutPrefixProvider *dpp, optional_yield y, rgw::lua::packages_t& failed_packages, std::string& output)
```
[the remark this answers is not in the document]
Comment (dang) [reply src/rgw/rgw_sal_rados.cc:3158]: RadosLuaManager is already below the line.  Once the pure virtual is in, maybe we can move this as a default implementation in StoreLuaManager.

File: src/rgw/rgw_obj_manifest.cc [hunk 3]
```diff
@@ -607,3 +607,50 @@ void RGWObjTier::dump(Formatter *f) const
...
+  if (!is_raw) {
+    rgw_raw_obj r;
+    rgw_obj_to_raw(zonegroup, zone_params, placement_rule, obj, &r);
+    return r;
+  }
+  return raw_obj;
+}
+
+// returns true on success, false on failure
+bool RGWRados::get_obj_data_pool(const rgw_placement_rule& placement_rule, const rgw_obj& obj, rgw_pool *pool)
```
Comment (dang) [inline src/rgw/rgw_obj_manifest.cc:652]: Why is a method from RGWRados here?

File: src/rgw/rgw_period.cc [hunk 4]
```diff
@@ -0,0 +1,682 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
Comment (dang) [inline src/rgw/rgw_period.cc:1]: Copyright?

**Review by dang — COMMENTED** [review]
Generally looks good.

File: src/rgw/rgw_sal_rados.cc [hunk 5]
```diff
@@ -3043,6 +3044,171 @@ int RadosLuaScriptManager::del(const DoutPrefixProvider* dpp, optional_yield y,
   return 0;
 }
```
Comment (yuvalif) [inline src/rgw/rgw_sal_rados.cc:3046]: why isn't this protected by `WITH_RADOSGW_LUA_PACKAGES` ?
